File: src/renderer/vulkan/FrameGraphResources.cpp
////////////////////////////////////////////////////////////////////////////////
/// Reaper
///
/// Copyright (c) 2015-2022 Thibault Schueller
/// This file is distributed under the MIT License
////////////////////////////////////////////////////////////////////////////////

#include "FrameGraphResources.h"

#include "Backend.h"
#include "common/ReaperRoot.h"
#include <profiling/Scope.h>

namespace Reaper
{
namespace
{
    void swap_resources(FrameGraphResources& resources)
    {
        std::swap(resources.textures, resources.textures_b);
        std::swap(resources.default_texture_views, resources.default_texture_views_b);
        std::swap(resources.additional_texture_views, resources.additional_texture_views_b);
    }
} // namespace

FrameGraphResources create_framegraph_resources(ReaperRoot& /*root*/, VulkanBackend& backend)
{
    FrameGraphResources resources = {};

    const VkEventCreateInfo event_info = {
        .sType = VK_STRUCTURE_TYPE_EVENT_CREATE_INFO,
        .pNext = nullptr,
        .flags = VK_EVENT_CREATE_DEVICE_ONLY_BIT,
    };

    for (auto& event : resources.events)
    {
        Assert(vkCreateEvent(backend.device, &event_info, nullptr, &event) == VK_SUCCESS);
    }

    // Volatile stuff is created later
    return resources;
}

void destroy_framegraph_resources(VulkanBackend& backend, FrameGraphResources& resources)
{
    destroy_framegraph_volatile_resources(backend, resources);

    for (auto& event : resources.events)
    {
        vkDestroyEvent(backend.device, event, nullptr);
    }
}

void allocate_framegraph_volatile_resources(ReaperRoot& root, VulkanBackend& backend, FrameGraphResources& resources,
                                            const FrameGraph::FrameGraph& framegraph)
{
    REAPER_PROFILE_SCOPE_FUNC();

    destroy_framegraph_volatile_resources(backend, resources); // FIXME Reuse previous resources here instead
    swap_resources(resources);

    using namespace FrameGraph;

    // Create buffers
    resources.buffers.resize(framegraph.BufferResources.size());

    for (u32 index = 0; index < framegraph.BufferResources.size(); index++)
    {
        const Resource& resource = framegraph.BufferResources[index];

        if (resource.is_used)
        {
            resources.buffers[index] = create_buffer(root, backend.device, resource.debug_name,
                                                     resource.properties.buffer, backend.vma_instance);
        }
        else
        {
            resources.buffers[index] = {};
        }
    }

    // Create textures
    resources.textures.resize(framegraph.TextureResources.size());
    resources.default_texture_views.resize(framegraph.TextureResources.size());

    for (u32 index = 0; index < framegraph.TextureResources.size(); index++)
    {
        const Resource& resource = framegraph.TextureResources[index];

        if (resource.is_used)
        {
            GPUTexture& new_texture = resources.textures[index];
            new_texture = create_image(root, backend.device, resource.debug_name, resource.properties.texture,
                                       backend.vma_instance);

            resources.default_texture_views[index] =
                create_image_view(root, backend.device, new_texture, resource.default_view.texture);
        }
        else
        {
            resources.textures[index] = {.handle = VK_NULL_HANDLE, .allocation = {}};
            resources.default_texture_views[index] = VK_NULL_HANDLE;
        }
    }

    // Create additional texture views
    resources.additional_texture_views.resize(framegraph.TextureViews.size());

    for (u32 index = 0; index < framegraph.ResourceUsages.size(); index++)
    {
        const ResourceUsage&      usage = framegraph.ResourceUsages[index];
        const ResourceHandle      resource_handle = usage.resource_handle;
        const ResourceViewHandles view_handles = usage.additional_views;

        nonstd::span<const GPUTextureView> texture_views_info =
            nonstd::make_span(framegraph.TextureViews.data() + view_handles.offset, view_handles.count);

        nonstd::span<VkImageView> texture_views =
            nonstd::make_span(resources.additional_texture_views.data() + view_handles.offset, view_handles.count);

        if (resource_handle.is_texture && usage.is_used)
        {
            const GPUTexture& texture = resources.textures[usage.resource_handle.index];

            for (u32 i = 0; i < texture_views.size(); i++)
            {
                texture_views[i] = create_image_view(root, backend.device, texture, texture_views_info[i]);
            }
        }
        else
        {
            for (u32 i = 0; i < texture_views.size(); i++)
            {
                texture_views[i] = VK_NULL_HANDLE;
            }
        }
    }
}

void destroy_framegraph_volatile_resources(VulkanBackend& backend, FrameGraphResources& resources)
{
    for (GPUBuffer& buffer : resources.buffers)
    {
        vmaDestroyBuffer(backend.vma_instance, buffer.handle, buffer.allocation);
        buffer = {};
    }

    for (GPUTexture& texture : resources.textures)
    {
        vmaDestroyImage(backend.vma_instance, texture.handle, texture.allocation);
        texture = {};
    }

    for (VkImageView& view : resources.default_texture_views)
    {
        vkDestroyImageView(backend.device, view, nullptr);
        view = VK_NULL_HANDLE;
    }

    for (VkImageView& view : resources.additional_texture_views)
    {
        vkDestroyImageView(backend.device, view, nullptr);
        view = VK_NULL_HANDLE;
    }
}

VkImage get_frame_graph_texture_handle(FrameGraphResources& resources, FrameGraph::ResourceHandle resource_handle)
{
    Assert(resource_handle.is_texture, "Wrong handle");

    const GPUTexture& texture = resources.textures[resource_handle.index];
    Assert(texture.handle != VK_NULL_HANDLE);

    return texture.handle;
}

FrameGraphTexture get_frame_graph_texture(FrameGraphResources& resources, const FrameGraph::FrameGraph& framegraph,
                                          FrameGraph::ResourceUsageHandle usage_handle)
{
    using namespace FrameGraph;

    const ResourceUsage& usage = GetResourceUsage(framegraph, usage_handle);
    const ResourceHandle resource_handle = usage.resource_handle;
    const Resource&      resource = GetResource(framegraph, resource_handle);
    Assert(resource_handle.is_texture, "Wrong handle");

    const GPUTexture&  texture = resources.textures[resource_handle.index];
    const VkImageView& default_view_handle = resources.default_texture_views[resource_handle.index];

    Assert(texture.handle != VK_NULL_HANDLE);
    Assert(default_view_handle != VK_NULL_HANDLE);

    nonstd::span<const VkImageView> additional_views = nonstd::make_span(
        resources.additional_texture_views.data() + usage.additional_views.offset, usage.additional_views.count);

    return FrameGraphTexture{
        .properties = resource.properties.texture,
        .view = resource.default_view.texture, // FIXME
        .handle = texture.handle,
        .default_view_handle = default_view_handle,
        .additional_views = additional_views,
        .image_layout = usage.access.image_layout,
    };
}

VkBuffer get_frame_graph_buffer_handle(FrameGraphResources& resources, FrameGraph::ResourceHandle resource_handle)
{
    Assert(!resource_handle.is_texture, "Wrong handle");

    const GPUBuffer& buffer = resources.buffers[resource_handle.index];
    Assert(buffer.handle != VK_NULL_HANDLE);

    return buffer.handle;
};

FrameGraphBuffer get_frame_graph_buffer(FrameGraphResources& resources, const FrameGraph::FrameGraph& framegraph,
                                        FrameGraph::ResourceUsageHandle usage_handle)
{
    using namespace FrameGraph;

    const ResourceUsage& usage = GetResourceUsage(framegraph, usage_handle);
    const ResourceHandle resource_handle = usage.resource_handle;
    const Resource&      resource = GetResource(framegraph, resource_handle);
    Assert(!resource_handle.is_texture, "Wrong handle");

    const GPUBuffer& buffer = resources.buffers[usage.resource_handle.index];

    return FrameGraphBuffer{
        // FIXME
        .properties = resource.properties.buffer,
        .default_view = resource.default_view.buffer,
        .handle = buffer.handle,
    };
}

} // namespace Reaper

File: src/renderer/vulkan/framegraph_resources.zig
// Port of src/renderer/vulkan/FrameGraphResources.{h,cpp}
//
// Owns the GPU objects behind the frame graph's declared resources. Everything
// except the event pool is volatile: it is torn down and rebuilt whenever the
// graph is rebuilt.

const std = @import("std");
const vk = @import("vulkan");

const buffer_module = @import("buffer.zig");
const fg = @import("../graph/frame_graph.zig");
const image_module = @import("image.zig");
const vma = @import("vma.zig").c;

const gpu_buffer = @import("../buffer/gpu_buffer.zig");
const gpu_texture_properties = @import("../texture/gpu_texture_properties.zig");
const gpu_texture_view = @import("../texture/gpu_texture_view.zig");

pub const FrameGraphTexture = struct {
    properties: gpu_texture_properties.GPUTextureProperties,
    default_view: gpu_texture_view.GPUTextureView,

    handle: vk.Image,
    default_view_handle: vk.ImageView,
    additional_views: []const vk.ImageView,
    image_layout: vk.ImageLayout,
};

pub const FrameGraphBuffer = struct {
    properties: gpu_buffer.GPUBufferProperties,
    default_view: gpu_buffer.GPUBufferView,

    handle: vk.Buffer,
};

pub const FrameGraphResources = struct {
    /// NOTE: you need as many events as concurrent synchronization barriers.
    /// For a first implementation, just have as many events as barriers.
    pub const event_count = 100;

    // Persistent
    events: [event_count]vk.Event,

    // Volatile
    buffers: std.ArrayList(buffer_module.GPUBuffer) = .empty,
    textures: std.ArrayList(image_module.GPUTexture) = .empty,
    default_texture_views: std.ArrayList(vk.ImageView) = .empty,
    additional_texture_views: std.ArrayList(vk.ImageView) = .empty,

    // NOTE: C++ keeps a second copy of the texture arrays purely so they can be
    // swapped at the top of allocate(). The destroy pass runs *before* that
    // swap and only touches the non-_b set, so the swap moves empty arrays
    // around and achieves nothing else. Left out here for that reason.

    allocator: std.mem.Allocator,

    pub fn init(vkd: anytype, device: vk.Device, allocator: std.mem.Allocator) !FrameGraphResources {
        const event_info = vk.EventCreateInfo{
            .s_type = .event_create_info,
            .p_next = null,
            .flags = .{ .device_only_bit = true },
        };

        var events: [event_count]vk.Event = undefined;
        var created: usize = 0;
        errdefer for (events[0..created]) |event| vkd.destroyEvent(device, event, null);

        while (created < event_count) : (created += 1) {
            events[created] = try vkd.createEvent(device, &event_info, null);
        }

        // Volatile stuff is created later
        return .{ .events = events, .allocator = allocator };
    }

    pub fn deinit(self: *FrameGraphResources, vkd: anytype, device: vk.Device, vma_instance: vma.VmaAllocator) void {
        self.destroyVolatileResources(vkd, device, vma_instance);

        self.buffers.deinit(self.allocator);
        self.textures.deinit(self.allocator);
        self.default_texture_views.deinit(self.allocator);
        self.additional_texture_views.deinit(self.allocator);

        for (self.events) |event| vkd.destroyEvent(device, event, null);
    }

    pub fn destroyVolatileResources(
        self: *FrameGraphResources,
        vkd: anytype,
        device: vk.Device,
        vma_instance: vma.VmaAllocator,
    ) void {
        for (self.buffers.items) |buffer| {
            if (buffer.handle != .null_handle) buffer_module.destroyBuffer(vma_instance, buffer);
        }
        self.buffers.clearRetainingCapacity();

        for (self.textures.items) |texture| {
            if (texture.handle != .null_handle) image_module.destroyImage(vma_instance, texture);
        }
        self.textures.clearRetainingCapacity();

        for (self.default_texture_views.items) |view| {
            if (view != .null_handle) vkd.destroyImageView(device, view, null);
        }
        self.default_texture_views.clearRetainingCapacity();

        for (self.additional_texture_views.items) |view| {
            if (view != .null_handle) vkd.destroyImageView(device, view, null);
        }
        self.additional_texture_views.clearRetainingCapacity();
    }

    /// FIXME Reuse previous resources here instead of destroying them.
    pub fn allocateVolatileResources(
        self: *FrameGraphResources,
        vkd: anytype,
        device: vk.Device,
        vma_instance: vma.VmaAllocator,
        framegraph: *const fg.FrameGraph,
    ) !void {
        self.destroyVolatileResources(vkd, device, vma_instance);

        const null_buffer = buffer_module.GPUBuffer{
            .handle = .null_handle,
            .allocation = null,
            .properties_deprecated = .{},
        };
        const null_texture = image_module.GPUTexture{ .handle = .null_handle, .allocation = null };

        // ---- Buffers ----
        try self.buffers.appendNTimes(self.allocator, null_buffer, framegraph.buffer_resources.items.len);

        for (framegraph.buffer_resources.items, 0..) |resource, index| {
            if (!resource.is_used) continue;

            self.buffers.items[index] = try buffer_module.createBuffer(
                vma_instance,
                resource.properties.buffer,
                .gpu_only,
            );
        }

        // ---- Textures ----
        try self.textures.appendNTimes(self.allocator, null_texture, framegraph.texture_resources.items.len);
        try self.default_texture_views.appendNTimes(
            self.allocator,
            .null_handle,
            framegraph.texture_resources.items.len,
        );

        for (framegraph.texture_resources.items, 0..) |resource, index| {
            if (!resource.is_used) continue;

            const new_texture = try image_module.createImage(vma_instance, resource.properties.texture);
            self.textures.items[index] = new_texture;

            // NOTE: C++ creates the default view unconditionally, which works
            // there only because every texture it declares happens to be
            // sampled or storage. A transfer-only texture cannot have a view at
            // all, so the usage is checked here.
            if (isViewable(resource.properties.texture.usage_flags)) {
                self.default_texture_views.items[index] = try image_module.createImageView(
                    vkd,
                    device,
                    new_texture.handle,
                    resource.default_view.texture,
                );
            }
        }

        // ---- Additional texture views ----
        try self.additional_texture_views.appendNTimes(
            self.allocator,
            .null_handle,
            framegraph.texture_views.items.len,
        );

        for (framegraph.resource_usages.items) |usage| {
            const view_handles = usage.additional_views;
            if (view_handles.count == 0) continue;

            if (!usage.resource_handle.is_texture or !usage.is_used) continue;

            const texture = self.textures.items[usage.resource_handle.index];
            const views_info = framegraph.texture_views.items[view_handles.offset..][0..view_handles.count];
            const views = self.additional_texture_views.items[view_handles.offset..][0..view_handles.count];

            for (views, views_info) |*view, view_info| {
                view.* = try image_module.createImageView(vkd, device, texture.handle, view_info);
            }
        }
    }

    /// vkCreateImageView rejects an image whose usage has no view-capable bit.
    fn isViewable(usage: gpu_texture_properties.GPUTextureUsage) bool {
        return usage.sampled or usage.storage or usage.color_attachment or
            usage.depth_stencil_attachment or usage.transient_attachment or usage.input_attachment;
    }

    pub fn getTextureHandle(self: *const FrameGraphResources, resource_handle: fg.ResourceHandle) vk.Image {
        std.debug.assert(resource_handle.is_texture);

        const texture = self.textures.items[resource_handle.index];
        std.debug.assert(texture.handle != .null_handle);

        return texture.handle;
    }

    pub fn getTexture(
        self: *const FrameGraphResources,
        framegraph: *const fg.FrameGraph,
        usage_handle: fg.ResourceUsageHandle,
    ) FrameGraphTexture {
        const usage = fg.getResourceUsage(framegraph, usage_handle);
        const resource = fg.getResource(framegraph, usage.resource_handle);
        const view_handles = usage.additional_views;

        return .{
            .properties = resource.properties.texture,
            .default_view = resource.default_view.texture,
            .handle = self.getTextureHandle(usage.resource_handle),
            .default_view_handle = self.default_texture_views.items[usage.resource_handle.index],
            .additional_views = self.additional_texture_views.items[view_handles.offset..][0..view_handles.count],
            .image_layout = usage.access.image_layout,
        };
    }

    pub fn getBufferHandle(self: *const FrameGraphResources, resource_handle: fg.ResourceHandle) vk.Buffer {
        std.debug.assert(!resource_handle.is_texture);

        const buffer = self.buffers.items[resource_handle.index];
        std.debug.assert(buffer.handle != .null_handle);

        return buffer.handle;
    }

    pub fn getBuffer(
        self: *const FrameGraphResources,
        framegraph: *const fg.FrameGraph,
        usage_handle: fg.ResourceUsageHandle,
    ) FrameGraphBuffer {
        const usage = fg.getResourceUsage(framegraph, usage_handle);
        const resource = fg.getResource(framegraph, usage.resource_handle);

        return .{
            .properties = resource.properties.buffer,
            .default_view = resource.default_view.buffer,
            .handle = self.getBufferHandle(usage.resource_handle),
        };
    }
};
